File: cs/aocutils.cs
using Spectre.Console;

namespace Shunty.AoC;

public static class AocUtils
{
    /// <summary>
    /// Output the given title text to the console in random ANSI
    /// colouring (if the shell supports it)
    /// </summary>
    public static void PrintTitle(string title)
    {
        var colours = new string[] { "red", "orange1", "yellow", "green", "blue", "purple", "violet" };
        var rnd = new Random();
        foreach(var c in $" * * * {title} * * * ")
        {
            var colour = colours[rnd.Next(colours.Length)];
            AnsiConsole.Markup($"[bold {colour}]{c}[/]");
        }
        AnsiConsole.WriteLine();
    }

    /// <summary>
    /// Try and find the input file for the given day.
    /// Search in the current directory, then in a './input/
    /// sub-directory i(f one exists) and then repeat in each
    /// parent directory tree for a few levels until we find
    /// an appropriate input file. If we reach the max number
    /// of levels then return an empty string.
    /// <para>
    /// This should make it easier to run the program from a number of different starting
    /// directories. eg from the solution (.sln) directory, from the project directory, or
    /// from the bin directory etc.
    /// </para>
    /// </summary>
    /// <param name="day">The day number of the input file we are looking for</param>
    /// <param name="suffix">An optional suffix to add to the file name to search for, eg: 'test'</param>
    /// <returns>The full file name of the input file, if found, otherwise an empty string</returns>
    public static string FindInputFile(int day, string suffix = "")
    {
        // An arbitrary number of levels to search up the directory tree
        const int maxParentLevels = 6;

        var dstart = Directory.GetCurrentDirectory();
        var dir = dstart;
        var dayfile = string.IsNullOrWhiteSpace(suffix)
            ? $"day{day:D2}-input"
            : $"day{day:D2}-input-{suffix}";
        int parentLevel = 0;
        while (parentLevel <= maxParentLevels)
        {
            // Look in this directory
            var fn = Path.Combine(dir, dayfile);
            if (File.Exists(fn))
            {
                return fn;
            }

            // Look in ./input sub-directory
            fn = Path.Combine(dir, "input", dayfile);
            if (File.Exists(fn))
            {
                return fn;
            }

            // Otherwise go up a directory
            var dinfo = Directory.GetParent(dir);
            if (dinfo == null)
            {
                break;
            }
            parentLevel++;
            dir = dinfo.FullName;
        }
        // Not found
        return "";
    }

}

public interface AocDaySolver
{
    int DayNumber { get; }
    Task Solve();
}

public static class AocDaySolverExtensions
{
    /// <summary>
    /// These extensions require the `Spectre.Console` Nuget package
    /// </summary>

    public static void ShowDayHeader(this AocDaySolver solver)
    {
        AnsiConsole.MarkupLine($"[bold]Day {solver.DayNumber}[/]");
    }

    public static void ShowDayResult<T>(this AocDaySolver _, int part, T solution, string suffix = "")
    {
        AnsiConsole.MarkupLine($"  [bold]Part {part}:[/] {solution?.ToString() ?? "<Unknown>"} {(string.IsNullOrWhiteSpace(suffix) ? "" : suffix)}");
    }

    public static void ShowDayResults<T>(this AocDaySolver _, T solution1, T solution2)
    {
        AnsiConsole.MarkupLine($"  [bold]Part 1:[/] {solution1?.ToString() ?? "<Unknown>"}");
        AnsiConsole.MarkupLine($"  [bold]Part 2:[/] {solution2?.ToString() ?? "<Unknown>"}");
    }

    public static void ShowDayResults<T1, T2>(this AocDaySolver _, T1 solution1, T2 solution2)
    {
        AnsiConsole.MarkupLine($"  [bold]Part 1:[/] {solution1?.ToString() ?? "<Unknown>"}");
        AnsiConsole.MarkupLine($"  [bold]Part 2:[/] {solution2?.ToString() ?? "<Unknown>"}");
    }
}

File: rs/src/aocutils.rs
use std::fmt::Display;
use std::fs;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

const MAX_SEARCH_LEVEL: i32 = 6;

/// Find the input file for the given day number by searching in the current
/// directory and then in a `./input` sub-directory, if it exists. If not found
/// then repeat the search up the parent directory chain for a few levels.
///
/// Returns the full path and filename if found and `None` if not found.
pub fn find_input_file(day_no: &i32) -> Option<String> {
    let inname = format!("day{:02}-input", day_no);
    // Start from the current directory
    let mut dir = ".".to_string();
    let mut fname = format!("{}/{}", dir, inname);

    let mut level = 0;
    loop {
        if level > MAX_SEARCH_LEVEL {
            break None;
        }

        // look in dir
        if Path::new(&fname).exists() {
            break Some(fname);
        }

        // look in dir/input
        fname = format!("{}/input/{}", dir, inname);
        if Path::new(&fname).exists() {
            break Some(fname);
        }

        // go up to the parent directory
        dir = format!("{}/..", dir);
        level += 1;
    }
}

/// Read all text from a file into a `String`. Remove all CR/LF.
pub fn get_text_from_file(filename: &str) -> io::Result<String> {
    let content = fs::read_to_string(filename)?;
    let result = content.replace(['\n', '\r'], "");
    Ok(result)
}

/// Read all text from a file into a `Vec<String>`
pub fn get_lines_from_file(filename: &str) -> io::Result<Vec<String>> {
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    reader.lines().collect()
}

/// Read a file with lines of integers separated by whitespace into a `Vec<Vec<i32>>`.
///
/// Each line of the input file is expected to have the format 'nn nnn n nn'
pub fn get_vec_of_vec_of_int_from_file(filename: &str) -> io::Result<Vec<Vec<i32>>> {

    let mut result: Vec<Vec<i32>> = Vec::new();

    match get_lines_from_file(filename) {
        Ok(lines) => {
            for (index, line) in lines.iter().enumerate() {
                let numbers: Result<Vec<i32>, _> = line
                    .split_whitespace()
                    .map(|part| part.parse::<i32>())
                    .collect();

                match numbers {
                    Ok(numbers) => {
                        result.push(numbers);
                    }
                    Err(e) => eprintln!("Error parsing line {}: {}", index + 1, e),
                }
            }
        }
        Err(e) => eprintln!("Error reading file: {}", e),
    }

    Ok(result)
}

/// Print the part number (usually 1 or 2) along with the solution result
pub fn print_day_result<T: Display>(part: &i32, result: T) {
    println!("  Part {}: {}", part, result);
}

